sqlglot/dialects/snowflake.py

```diff
@@ -33,6 +33,21 @@ def _check_int(s: str) -> bool:
     return s.isdigit()
 
 
+def _parse_to_array(args: t.List) -> exp.Expression:
+    arg = seq_get(args, 0)
+    if isinstance(arg, exp.Expression):
+        from sqlglot.optimizer.annotate_types import annotate_types
+
+        # https://docs.snowflake.com/en/sql-reference/functions/to_array
+        arg = annotate_types(arg)
+        if arg.is_type(exp.DataType.Type.ARRAY):
+            return arg
+        if arg.is_type(exp.DataType.Type.VARIANT):
+            return exp.Anonymous(this="TO_ARRAY", expressions=[arg])
```

Comment:
@wezham addressed the `VARIANT` case more conservatively here because I realized it's more complex. The difficulty is that we need to check the type of the _wrapped_ value of the variant, so I think we'd only be able to do that if we checked a few known cases like casts to variants etc where we can access the wrapped value. Didn't seem worth adding that complexity..